### PR TITLE
Update documentation to new TestListeners

### LINKED ISSUE #535: Add documentation to Listeners
The documentation on `TestListeners` must be updated to the new version. Some functions are not there, and some functions changes names. For example, `afterSpecClass`

### PR DESCRIPTION
**Pull request by Kerooker**
This commit updates the documentation on TestListeners, to include the new test listeners (such as `beforeSpecClass` and `afterSpecClass`), while also adapting the old listeners to the new ones (the old ones are deprecated as of v3.2.0).

A small detail was also updated on the documentation, to include AnnotationSpec in the testing styles.

Fixes #535

### CODE REVIEW

**Comment by Kerooker** [comment]
Please, take a look and see if I missed a point or two about this. I don't know if I read all the docs correctly :x

File: doc/reference.md [hunk 1]
```diff
@@ -321,6 +323,19 @@ class MyTestClass : WordSpec() {
 }
 ```
 
+It's also important to notice that every `Spec` is also a `TestListener`, therefore you may override these functions directly in `Spec`.
```
Comment (sksamuel) [inline doc/reference.md]: important to note that

File: doc/reference.md [hunk 2]
```diff
@@ -329,10 +344,12 @@ The full list of the functions in the `TestListener` interface is as follows:
 
 |Function|Purpose|
 |--------|-------|
-|beforeTest|This function will be invoked each time a new Test Case is executed.|
-|afterTest|Is invoked when a Test Case has finished. This includes when a test case is ignored (skipped), passes (is successful), or fails (errors).|
+|beforeTest|Is invoked each time a new Test Case is executed. If it's marked as Ignored, this won't execute.|
```
Comment (sksamuel) [inline doc/reference.md]: If the test is marked...

File: doc/reference.md [hunk 3]
```diff
@@ -329,10 +344,12 @@ The full list of the functions in the `TestListener` interface is as follows:
...
+|afterTest|Is invoked when a Test Case has finished. This includes when a test case is passes (is successful), or fails (errors).|
 |beforeSpec|Is invoked each time a Spec is started, before any `beforeTest` functions are invoked. |
 |afterSpec|Is invoked each time a Spec completes, after all `afterTest` functions are invoked. |
+|beforeSpecClass|Is invoked when the engine is preparing the spec to be executed. It will be executed only once, regardless of how many times the [Spec is instantiated](isolation_mode.md)
```
Comment (sksamuel) [inline doc/reference.md]: nice link through to isolation mode

**Pull request merged**